### PR TITLE
[SYCL] Prevent unintended sign extensions

### PR DESCRIPTION
**Pull request by ianayl**
This is to resolve a Coverity hit, pointing out that a potentially unintended sign extension is possible:

- `SectionHeaderNum` and `SectionHeaderSize` are unsigned i16.
- By C++ promotion rules, `SectionHeaderNum * SectionHeaderSize` results in a signed i32 value.
- However, `SectionHeaderOffset` is i64, so the i32 product is promoted to i64 via sign extension
- If `SectionHeaderNum * SectionHeaderSize` results in a number greater than 0x7FFFFF (greatest signed i32 number), then the upper bits get set to 1, producing the wrong result.

This PR adds a cast to ensure the promotions do not result in unexpected sign extensions: While I recognize that the odds of `SectionHeaderNum` and `SectionHeaderSize` getting as big as sqrt(0x7FFFFFF) is rather unlikely, this change does not hurt or slow down the code, and is probably the safer thing to do anyway. If it is decided that this change is not necessarily, feel free to close the PR, and I will fix the Coverity triage to reflect the decision.

### CODE REVIEW

**Comment by ianayl** [comment]
@intel/unified-runtime-reviewers Friendly ping, thanks in advance!

File: sycl/source/detail/ur.cpp [hunk 1]
```diff
@@ -300,7 +300,8 @@ static bool checkELFSectionPresent(const std::string &ExpectedSectionName,
 
   // End early if we do not have the expected number of section headers or
   // if the read section string header index is out-of-range.
-  if (ImgSize < SectionHeaderOffset + SectionHeaderNum * SectionHeaderSize ||
+  if (ImgSize < SectionHeaderOffset +
+                    (uint64_t)SectionHeaderNum * SectionHeaderSize ||
```
Comment (aarongreig) [inline sycl/source/detail/ur.cpp:304]: can we make this a static_cast instead?

Comment (ianayl) [reply sycl/source/detail/ur.cpp]: I've gone ahead and used static cast instead, although may I ask why? I am assuming it is because static_cast is safer.

Comment (aarongreig) [reply sycl/source/detail/ur.cpp]: yeah it's safer and it's also in the LLVM coding standards: [prefer c++ style casts](https://llvm.org/docs/CodingStandards.html#prefer-c-style-casts)

**Review by aarongreig — APPROVED** [review]

**Comment by ianayl** [comment]
Thanks Aaron for the review!

@intel/llvm-gatekeepers This is ready for merging, thanks!

**Pull request merged**